pkg/domain/domain.go

```diff
@@ -1808,6 +1810,28 @@ func (do *Domain) GetPDHTTPClient() pdhttp.Client {
 	return nil
 }
 
+func decodePrivilegeEvent(resp clientv3.WatchResponse) PrivilegeEvent {
+	var msg PrivilegeEvent
+	for _, event := range resp.Events {
+		if event.Kv != nil {
+			val := event.Kv.Value
+			if len(val) > 0 {
+				err := json.Unmarshal(val, &msg)
+				if err == nil {
+					break
+				}
+				logutil.BgLogger().Warn("decodePrivilegeEvent unmarshal fail", zap.Error(err))
+			}
+		}
+	}
+	// In case something is wrong, for example, old version tidb mixed with newer, the unmarshal would fail.
+	// Then we fallback to the old way: reload all the users.
+	if len(msg.UserList) == 0 {
+		msg.All = true
+	}
```

Comment:
This differs from the old logic.   @D3Hunter 
What happen if something is wrong? if val does have data, but the data is not what we expected?
Neither the UserList contains data nor the ALL flag is set.
Our expected case is reload all on error to tolerance the cases, but the suggested change here is ignore on error.

Reply:
> What happen if something is wrong? if val does have data, but the data is not what we expected?

if no third-part involved, older version tidb write a `empty value`, new version write a `valid json`, so when will the data be invalid? If it's caused by bug, then let's fix it, else I think the event consumer should do as what the producer told it to do. and maybe you can add a `intest.Assert` to check that either `All=true` or the `len(userList) > 0`, the producer side should make sure no such invalid data are sent to ETCD

and as you have mentioned, the event is not that critical, i think it's acceptable

Reply:
> > What happen if something is wrong? if val does have data, but the data is not what we expected?
> 
> if no third-part involved, older version tidb write a `empty value`, new version write a `valid json`, so when will the data be invalid? If it's caused by bug, then let's fix it, else I think the event consumer should do as what the producer told it to do. and maybe you can add a `intest.Assert` to check that either `All=true` or the `len(userList) > 0`, the producer side should make sure no such invalid data are sent to ETCD
> 
> and as you have mentioned, the event is not that critical, i think it's acceptable

I agree that we should ignore the message and skip to update privilege cache when failed to decode the json and only handle the case `val == ""`. Because we also have a loop to guarantee the cache should be updated in a certain time, it is not a big deal.

We can also generate some warnings logs when meet a invalid json.